### DISCUSSION TITLE
Can we have default properties, and base web interfaces too inside the interfaces?

### DISCUSSION THREAD

**Discussion by Nefcanto** [dotnet/csharplang#8075] [Q&A]
Default methods inside interfaces provide a neat way for composition. Since C# does not have multiple inheritance, we can extract common functionality into separate interfaces and add them to existing classes easily.

But they are very limited. Two points:

1. A lot of time we need type-level state management (fields and properties) to be used amongst multiple methods.
2. We can't use them for the web. We don't have access to `HttpContext` or `return Ok` in our default methods.

Can these two new features be added to the C# language? Can we have `ISeoController` which has a default property called `PathSeo` and two default methods using that property, and also they access the `HttpContext`?

**Comment by HaloFour** [comment]
You can define properties on interfaces, including with default implementations.  However, interfaces cannot contain state.  They have no storage for it, and attempting to support that would bring with it all of the complexity (and problems) of multiple inheritance.

As for the web, that sounds like a question for ASP.NET or whatever framework you use.  There's nothing about the language that forbids their use in such cases.

**Reply by colejohnson66** [reply]
Would the proposed `field` keyword on properties work on DIMs?

**Reply by HaloFour** [reply]
No, that would require that the compiler can emit a backing field for the property, and interfaces can't have fields.

**Reply by Nefcanto** [reply]
@HaloFour, I wish C# would support composition the way PHP or JS supports it. Composition in C# is a true pain. You need to constantly repeat yourself in boilerplate code. You define a common method, but then you should create wrapper methods in every other class to expose that composed common method:

```
public class Composable
{
      public void DoSomething()
      {
      }
}

public class FirstComposing
{
      private Composable composable; // boilerplate, painful
      
      public void DoSomething() // boilerplate, painful
      {
             composable.DoSomething(); // boilerplate, painful
      }
}

public class SecondComposing
{
      private Composable composable; // boilerplate, painful
      
      public void DoSomething() // boilerplate, painful
      {
             composable.DoSomething(); // boilerplate, painful
      }
}
```

Default methods in interfaces are not much of a help. They are very useless when it comes to real-world problems. And the .NET team does not listen to the community. We have different dimensions like internationalization, CQRS, multitenancy, hierarchical structures, modularization, ... and we can't put all of them in one base controller. We need to be able to compose from amongst existing controller actions. I need to be able to create a controller, and with minimal code (minimal boilerplate) compose it to be internationalized, multitenant, support CQRS to be read-only or read-write, support security, use different modules in it, etc.

@HaloFour, are you a member of the .NET team? If so please here the voice of the community. We are in pain here.

**Reply by theunrepentantgeek** [reply]
> And the .NET team does not listen to the community.

This is patently untrue. There is plenty of evidence that they do listen, including features that have been redesigned or even abandoned.

**Reply by CyrusNajmabadi** [reply]
Not on that, we've shipped several features design and implemented *by* community members :)

**Reply by Nefcanto** [reply]
@theunrepentantgeek, but they didn't listen to our pain: https://github.com/dotnet/csharplang/discussions/7406

We have created an SPL (software product line) based on docker, React, Qwik, MariaDB, and for our API we're using .NET. It also has a small headless CMS build into it. We tolerate a lot of pain because of the lack of multiple inheritance or simple composition (without tremendous boilerplate).

This is part of our React code to create a list:

```
const Customers = () => {
   return <List
        entityType="Customer"
        hasSearch
   />
}
````

A simple `hasSearch` boolean attribute/prop, and boom. A search would appear on the list like magic. With full functionality and tons of features behind the scene. Now we want to make our API's searchable: What should we do? Do we have this syntax?

```
public class CustomerController : ReadWriteController<Customer>
{
    #HasSearch
}
```

No. We don't. Do we have multiple inheritance? Nope. Do we have stateful generic interfaces with access to `HttpContext`? No. What can we do? How about this code?

```
public class CustomerController : ReadWirteController<Customer>, ISearchable
{
    [HttpGet]
    public IActionResult Search(string term)
    {
         // logic for search
        return Ok();
    }
}
```

Inefficient? Yes. Because now I'm forced to create tons of boilerplate spread over many APIs. And what if I want to change the behavior? Let's say I want to add a `FastSearch` to it. Or I just decided that the term `Search` is not a good action name and I want to change it to `PerformSearch`? What if I have 500 APIs over 50 projects? See the pain?

How about this code?

```
public class CustomerController : ReadWriteController<Customer>
{
    private SearchService searchService;
   
    public CustomerController(ISearchService searchService)
    {
       this.searchService = searchService;
    }

    [HttpGet]
    public IActionResult Search(string term)
    {
        return searchService.Search(term);
    }
}
```

Still inefficient and has every problem that comes with the interface.

**Reply by davidfowl** [reply]
You can build this without any language features. You don't need to use controllers, nor do you need to model them with the "mixin like" pattern and default interface methods.

**Reply by Nefcanto** [reply]
@davidfowl, that's interesting. I can't wait to see an example. We've been struggling with this for years.

**Reply by davidfowl** [reply]
I mentioned to you on the other thread that you can use code generation as a strategy to reduce "typing" the boilerplate.

Help us understand which part of the composition problem you are trying to fix. 

> What if I have 500 APIs over 50 projects? See the pain?

Are you saying you don't want to update 500 APIs? You're avoiding that by putting it on the base class? Can you flesh out this search sample?

**Reply by CyrusNajmabadi** [reply]
> @theunrepentantgeek, but they didn't listen to our pain: https://github.com/dotnet/csharplang/discussions/7406

People listened and responded to you.  You seemed to stop responding to them.  It's unclear how you're interpreting this as us not listening.

**Reply by CyrusNajmabadi** [reply]
> Because now I'm forced to create tons of boilerplate spread over many APIs. 

Why not use code-generation here?  Making it so that the compiler writes the boilerplate for you is a prime motivating reason for their existence :)

**Reply by Nefcanto** [reply]
@davidfowl, yes you're right. We can put it in the base class, **as long as we only have one base class**. That's the point. When it comes to complexity, a single base class becomes God object. At least 2K lines of code with is a mess. So, we don't have one base class. We have many base classes. This is our real base classes right now:

```
AudioController
AuthenticatedController
AuthorizedController
CacheController
ClientLookupReadController
ClientLookupReadWriteController
FileController
ImageController
PublicController
ReadController
ReadTreeController
ReadWriteController
SearchController
ReadWriteTreeController
VideoController
```

It's a mess now. We just tolerate the pain hopping to find a solution.

Let's say a new customer comes to us. He wants to sell flowers. Our developer needs to create a module for him called `Flowers`. When it comes to API, the APIs we deliver should be able to:

- Searchable
- Read/Write
- Stream videos
- Save/serve images
- Public for the site, and private (authenticated and authorized) for the admin panel
- Cacheable for performance

Right now what do we do? We create a `FlowersController` in our admin, we inherit from the read/write controller (to give us CRUD operations), and then we have to repeat ourselves by writing a lot of boilerplate code to reuse search, streaming, caching, etc.

The point here is the language support (besides code generation). Consider this syntax:

```
public class FlowersController : ReadWriteController<Flower>
{
    #Searchable<Flower>
    #Cacheable<Flower>
    #Streamable<Flower>
}
```

And imagine that the compiler finds this code, and puts it in that place:

```
public composable Searchable<T> where T class, new()
{
    [HttpGet]
    public IActionResult Search(string term)
    {
        // centralized generic logic
        return JsonResult(someResult);
    }
}
```

The `#` keyword which indicates that something should be put here, and the `composable` keyword that indicates this code is not executable directly, and its content should be put inside a place literally.

Wouldn't that be great?

**Reply by CyrusNajmabadi** [reply]
> The # keyword which indicates that something should be put here,

Why can you not do this with attributes and source generators?  You'd use a syntax like: 

```c#
[Searchable<Flower>()]
[Cacheable<Flower>()]
[Streamable<Flower>()]
public partial class FlowersController ...
```

And the generator will spit up whatever simple boilerplate is needed to hook these disparate systems together?

**Reply by Nefcanto** [reply]
@CyrusNajmabadi, is there an example of that? I mean the syntax is very good.

**Reply by Nefcanto** [reply]
@CyrusNajmabadi, please guide me on how can I use attributes on adding methods and properties via code generation.

**Reply by davidfowl** [reply]
You would define your own attributes and build a source generator:
https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview

Your controllers would all need to be partial and you would emit another partial that would code spit the implementation based on the attribute. This isn't a trivial thing to do but it is pretty reasonable given your constraints.

The other approach would be to extend the MVC framework itself to detect those attributes and inject more action methods into the model. You can use application model conventions to customize the framework https://www.stevejgordon.co.uk/customising-asp-net-mvc-core-behaviour-with-an-iapplicationmodelconvention. 

Again, non-trivial but doable given your constraints.

**Reply by CyrusNajmabadi** [reply]
> Again, non-trivial but doable given your constraints.

Given taht there are 500 APIs over 50 projects, the investment here would certainly be worthwhile.

**Reply by Nefcanto** [reply]
@davidfowl, thank you for your help and guidance. I will test them and come back with the results.

**Comment by huoyaoyuan** [comment]
Multiple inheritance is just one way to compose code. I'd like to explore better and friendly way to compose codes in more dimensions.
For example, currently in CoreLib, we need to declare each functionality for every numeric types. Composing may be used to reduce the whole matrix.
